File: finite_automaton/domol_automat.cpp
#include "domol_automat.h"

domol_automat::domol_automat() : Automat()
{

}

domol_automat::domol_automat(std::vector<std::string> alph, std::vector<std::vector<int>> st_transition, int start_st, std::vector<int> acc_sts, std::vector<bool> backup, std::vector<bool> read) : Automat({'c'}, st_transition, start_st, acc_sts)
{
    this->alphabet = alph;
    this->backup = backup;
    this->read = read;
}

bool domol_automat::validate(std::string s)
{
    state = start_state;
    for(int i = 0; i < s.length();)
    {
        try
        {
            int alph;
            if(backup[state])
            {
                i--;
            }
            if(read[state])
            {
                std::string Type;
                if(isdigit(s[i]))
                {
                    Type = "number";
                }
                else
                {
                    switch(s[i])
                    {
                        case '{':
                            Type = "{";
                            break;
                        case '}':
                            Type = "}";
                            break;
                        case '(':
                            Type = "(";
                            break;
                        case ')':
                            Type = ")";
                            break;
                        case '*':
                            Type = "*";
                            break;
                        case ';':
                            Type = ";";
                            break;
                        case '=':
                            Type = "=";
                            break;
                        case '<':
                            Type = "<";
                            break;
                        case '>':
                            Type = ">";
                            break;
                        case ' ':
                            Type = "space";
                            break;
                        case '$':
                            Type = "$";
                            break;
                        default:
                            if(s[i] == 'a' || s[i] == 'á' ||  s[i] == 'b' ||  s[i] == 'c' ||  s[i] == 'd' ||  s[i] == 'e' || s[i] == 'é' ||  s[i] == 'f' ||  s[i] == 'g' ||  s[i] == 'h' ||  s[i] == 'i' || s[i] == 'í' || s[i] == 'j' ||  s[i] == 'k' ||  s[i] == 'l' ||  s[i] == 'm' ||  s[i] == 'n' ||  s[i] == 'o' || s[i] == 'ó' || s[i] == 'ö' || s[i] == 'ő' || s[i] == 'p' ||  s[i] == 'q' ||  s[i] == 'r' ||  s[i] == 's' ||  s[i] == 't' ||  s[i] == 'u' || s[i] == 'ú' || s[i] == 'ü' || s[i] == 'ű' ||  s[i] == 'v' ||  s[i] == 'w' ||  s[i] == 'x' ||  s[i] == 'y' ||  s[i] == 'z' ||  s[i] == 'A' || s[i] == 'Á' || s[i] == 'B' ||  s[i] == 'C' ||  s[i] == 'D' ||  s[i] == 'E' || s[i] == 'É' || s[i] == 'F' ||  s[i] == 'G' ||  s[i] == 'H' ||  s[i] == 'I' || s[i] == 'Í' ||  s[i] == 'J' ||  s[i] == 'K' ||  s[i] == 'L' ||  s[i] == 'M' ||  s[i] == 'N' ||  s[i] == 'O' || s[i] == 'Ó' || s[i] == 'Ö' || s[i] == 'Ő' || s[i] == 'P' ||  s[i] == 'Q' ||  s[i] == 'R' ||  s[i] == 'S' ||  s[i] == 'T' ||  s[i] == 'U' || s[i] == 'Ú' || s[i] == 'Ü' || s[i] == 'Ű' || s[i] == 'V' ||  s[i] == 'W' ||  s[i] == 'X' ||  s[i] == 'Y' ||  s[i] == 'Z')
                            {
                                Type = "letter";
                            }
                            else
                                Type = "other";
                    }
                }
                alph = findAlph(Type);
            }
            state = state_transitions[state][alph];
            if(state == 0)
            {
                return false;
            }
        }
        catch(const std::string e)
        {
            return false;
        }
    }
    return isInAcceptingState();
}

int domol_automat::findAlph(std::string s)
{
    for(int i = 0; i < alphabet.size(); i++)
    {
        if(alphabet[i] == s)
        {
            return i;
        }
    }
    throw std::string("Item Not Found In Vector.");
}

File: finite_automaton/domol_automat.py
from automat import *
import string
class DomolAutomat(Automat):
    def __init__(self, alphabet, state_transitions, start_state, accepting_states, backup, read):
        #self.alphabet = alphabet
        #self.state_transitions = state_transitions
        #self.start_state = start_state
        #self.accepting_states = accepting_states
        super().__init__(alphabet, state_transitions, start_state, accepting_states)
        self.backup = backup
        self.read = read 

    def validate(self, s):
        self.state = self.start_state
        i = 0
        while i < len(s):
            try:
                alph = ""
                Type = "other"
                if self.backup[self.state]:
                    i -= 1
                elif self.read[self.state]:
                    if s[i] in string.digits:
                        Type = "number"
                    elif s[i] in string.ascii_letters or s[i] in "áéíóöőúüűÁÉÍÓÖŐÚÜŰ":
                        Type = "letter"
                    elif s[i] == "{":
                        Type = "{"
                    elif s[i] == "}":
                        Type = "{"
                    elif s[i] == "(":
                        Type = "("
                    elif s[i] == "*":
                        Type = "*"
                    elif s[i] == ")":
                        Type = ")"
                    elif s[i] == ":":
                        Type = ":"
                    elif s[i] == "=":
                        Type = "="
                    elif s[i] == "<":
                        Type = "<"
                    elif s[i] == ">":
                        Type = ">"
                    elif s[i] == " ":
                        Type = "space"
                    elif s[i] == "$":
                        Type = "$"
                    else:
                        Type = "other"
                    i += 1

                alph = self.findAlph(Type)   
                self.state = self.state_transitions[self.state][alph]
                if self.state == 0:
                    return False
            except ItemNotFound:
                return False
        return self.isInAcceptingState()
